### PR TITLE
8269122: The use of "extern const" for Register definitions generates poor code

### PR DESCRIPTION
**Pull request by theRealAph**
Register definitions in HotSpot are declared as "extern const"
for ancient-historical reasons. We should stop doing that: it would
make the assembler significantly faster and smaller, improving both
bootstrap time and compilation speed.

This change shaves 2% off the size of the text section of libjvm.so on AArch64.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269122](https://bugs.openjdk.java.net/browse/JDK-8269122): The use of "extern const" for Register definitions generates poor code

### Reviewers
 * [Andrew Dinn](https://openjdk.java.net/census#adinn) (@adinn - **Reviewer**) ⚠️ Review applies to 0f6396c09d4a781583e30855001533022fae6ce9
 * [Kim Barrett](https://openjdk.java.net/census#kbarrett) (@kimbarrett - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 621c25be92acd73984d37d952af3442e2159488f

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4554/head:pull/4554` \
`$ git checkout pull/4554`

Update a local copy of the PR: \
`$ git checkout pull/4554` \
`$ git pull https://git.openjdk.java.net/jdk pull/4554/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4554`

View PR using the GUI difftool: \
`$ git pr show -t 4554`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4554.diff">https://git.openjdk.java.net/jdk/pull/4554.diff</a>

</details>

### CODE REVIEW

**Review by adinn — APPROVED** [review]
Looks ok so long as you update the copyright

**Comment by theRealAph** [comment]
> Looks ok so long as you update the copyright

Done.

File: src/hotspot/share/asm/register.hpp [hunk 1]
```diff
@@ -45,8 +45,29 @@ class AbstractRegisterImpl {
   int value() const                              { return (int)(intx)this; }
 };
 
+#define AS_REGISTER(type,name)         ((type)name##_##type##EnumValue)
+
+#ifndef USE_EXTERN_CONST_FOR_REGISTER_DEFINITIONS
```
Comment (kimbarrett) [inline src/hotspot/share/asm/register.hpp:50]: I can see this conditionalization being useful for development and testing of this change, but why keep it and the old, now dead, code?

**Review by kimbarrett — CHANGES_REQUESTED** [review]

**Review by vnkozlov — APPROVED** [review]
Looks good. Let me test it in our environment. I will let you know results.

**Comment by vnkozlov** [comment]
Tier1 testing is good.

File: src/hotspot/share/asm/register.hpp (hunk 1 above)
Comment (theRealAph) [reply src/hotspot/share/asm/register.hpp:50]: The old code was a workaround for compilers that don't handle inline constants well. There are ports that I don't know about or can't test, and there are some outside the main jdk/jdk tree. If this change breaks any of them, all they have to do is define USE_EXTERN_CONST_FOR_REGISTER_DEFINITIONS. If it turns out that no such systems still exist, we can take it out.

Comment (theRealAph) [reply src/hotspot/share/asm/register.hpp:50]: Are you satisfied with that, Kim? Thanks.

Comment (kimbarrett) [reply src/hotspot/share/asm/register.hpp:50]: Not really.  I don't think we should be retaining dead code to support potential / imaginary compilers.  If such a compiler still exists (and is sufficiently modern to support C++14), the relevant code can be retrieved from the repository.

Comment (theRealAph) [reply src/hotspot/share/asm/register.hpp]: OK, I need to be braver! Done. :-)

**Review by kimbarrett — APPROVED** [review]
This looks good.

I think there are further cleanups to do, though they can be done in
followup RFEs.

(1) With REGISTER_DEFINITION now empty, it and uses of it can be removed.

(2) While looking at this I stumbled across the DONT_USE_REGISTER_DEFINES
blocks for s390 and ppc. Those are defining shadowing macros for the various
register constants, which is pointless after the changes being made here.
And I think these are very nearly the only use of the xxxEnumValue
enumerators; on a quick scan I spotted only two others that seem like they
could easily be fixed.

**Comment by theRealAph** [comment]
> (2) While looking at this I stumbled across the DONT_USE_REGISTER_DEFINES
> blocks for s390 and ppc. Those are defining shadowing macros for the various
> register constants, which is pointless after the changes being made here.
> And I think these are very nearly the only use of the xxxEnumValue
> enumerators; on a quick scan I spotted only two others that seem like they
> could easily be fixed.

Indeed. My plan for now is to ask the maintainers of those platforms to look at it.
If they don't want to do that, I'll do it. I think the automated testing should be enough.

**Pull request closed without merging**